### PR TITLE
[DON'T MERGE] LITE-15627 Message retrying

### PR DESCRIPTION
**Pull request by net-free**
Merge conditions:
- Base retrying with delay (LITE-18052, on review)
- Dead queue (LITE-18053, open)
- Documentation (LITE-18056, open)

Optional:
- Delay queue limits (LITE-18054, open)
- Dead queue commands (LITE-18055, open)

### CODE REVIEW

File: dj_cqrs/delay.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        self.delivery_tag = delivery_tag
+        self.payload = payload
+        self.eta = utc_now() + timedelta(seconds=delay)
+
+
+class DelayQueue:
+    """Queue for delay messages."""
+
+    def __init__(self):
+        self._queue = PriorityQueue()
```
Comment (maxipavlovic) [inline dj_cqrs/delay.py:29]: Just asking: in Python docs there is this: https://docs.python.org/3/library/queue.html#queue.PriorityQueue

```python
The lowest valued entries are retrieved first (the lowest valued entry is the one returned by sorted(list(entries))[0]). A typical pattern for entries is a tuple in the form: (priority_number, data).

If the data elements are not comparable, the data can be wrapped in a class that ignores the data item and only compares the priority number:

from dataclasses import dataclass, field
from typing import Any

@dataclass(order=True)
class PrioritizedItem:
    priority: int
    item: Any=field(compare=False)
```

Don't we need to wrap?

File: dj_cqrs/mixins.py [hunk 2]
```diff
@@ -404,3 +408,54 @@ def cqrs_delete(cls, master_data):
             raise NotImplementedError
 
         return cls.cqrs.delete_instance(master_data)
+
+    @staticmethod
```
Comment (maxipavlovic) [inline dj_cqrs/mixins.py:412]: This two methods seem to have no association with instance. Do you think it's better to move implementation to manager or to leave here?

File: dj_cqrs/mixins.py [hunk 3]
```diff
@@ -404,3 +408,54 @@ def cqrs_delete(cls, master_data):
...
+        min_value = 0
+        if not isinstance(max_retries, int) or max_retries < min_value:
+            logger.warning(
+                "Settings max_retries={} is invalid, using default {}".format(
+                    max_retries, default_max_retries,
+                )
+            )
+            max_retries = default_max_retries
+
+        return current_retry < max_retries
```
Comment (maxipavlovic) [inline dj_cqrs/mixins.py:438]: Verify, please, that we need strict comparison here.

File: dj_cqrs/mixins.py [hunk 4]
```diff
@@ -404,3 +408,54 @@ def cqrs_delete(cls, master_data):
...
+        """Checks if we should retry the message after current attempt.
+
+        :param current_retry: Current number of message retries.
+        :type current_retry: int
+        :param exception: Exception instance raised during message consume.
+        :type exception: Exception, optional
+        :return: True if message should be retried, False otherwise.
+        :rtype: bool
+        """
+        default_max_retries = 10
```
Comment (maxipavlovic) [inline dj_cqrs/mixins.py:423]: There is a file: `dj_cqrs/constants.py`
Please, consider to place `default_max_retries` and `default_retry_delay ` (below) constants there.

File: dj_cqrs/signals.py [hunk 5]
```diff
@@ -73,6 +73,7 @@ def post_save(cls, sender, **kwargs):
                 instance.pk,
                 queue,
                 previous_data,
+                expires=get_expires_datetime(),
```
Comment (maxipavlovic) [inline dj_cqrs/signals.py:76]: I remember that we discussed that expiration may be set per model(instance), you will change that later?

File: dj_cqrs/transport/rabbit_mq.py [hunk 6]
```diff
@@ -262,8 +294,23 @@ def _get_common_settings(cls):
     @staticmethod
     def _get_consumer_settings():
         queue_name = settings.CQRS['queue']
-        consumer_prefetch_count = settings.CQRS.get('consumer_prefetch_count', 10)
+        if 'consumer_prefetch_count' in settings.CQRS:
```
Comment (maxipavlovic) [inline dj_cqrs/transport/rabbit_mq.py:297]: That reminds us that we need to release version 2.0 :)

File: dj_cqrs/utils.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+logger = logging.getLogger('django-cqrs')
+
+
+def get_expires_datetime():
+    """Calculates when message should expire.
+
+    :return: datetime instance, None if infinite
+    :rtype: datetime
+    """
+    default_message_ttl = 86400  # 1 day
```
Comment (maxipavlovic) [inline dj_cqrs/utils.py:18]: Constants?

File: dj_cqrs/utils.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import logging
+from datetime import datetime, timedelta, timezone
+
+from django.conf import settings
+
+
+logger = logging.getLogger('django-cqrs')
+
+
+def get_expires_datetime():
```
Comment (maxipavlovic) [inline dj_cqrs/utils.py:12]: I don't know what's really better here: to have one result type (return unreachable DT from future) or to do as you've done. What do you think?

File: requirements/dev.txt [hunk 9]
```diff
@@ -3,3 +3,4 @@ pika>=1.0.0
 kombu==4.6.*
 ujson==3.0.0
 django-model-utils==4.0.0
+python-dateutil >= 2.4
```
Comment (maxipavlovic) [inline requirements/dev.txt:6]: Could you explain, why did you add this library?

**Review by maxipavlovic — CHANGES_REQUESTED** [review]
General comment: please, update copyright dates in all the files, that you've changed :)

File: dj_cqrs/dataclasses.py [hunk 10]
```diff
@@ -45,6 +54,37 @@ def __init__(
         else:
             self.__correlation_id = get_correlation_id(signal_type, cqrs_id, instance_pk, queue)
 
+        self.__expires = expires
+        self.__retries = retries
+
+    @classmethod
+    def from_message(cls, dct):
```
[the remark this answers is not in the document]
Comment (net-free) [reply dj_cqrs/dataclasses.py:61]: At this case for me it was important to point at "message" as a source (from where this dct is).
I suggest to store there additional logic for backward compatibility (older broker messages formats) and other payload building logic.

File: dj_cqrs/mixins.py (hunk 2 above)
Comment (net-free) [reply dj_cqrs/mixins.py:412]: I think for end library users it's better to stay here because it's easy to:
- implement
- understand
- support backward compatibility

File: dj_cqrs/signals.py (hunk 5 above)
Comment (net-free) [reply dj_cqrs/signals.py:76]: Expiration is per Service (maybe there was a misunderstanding).
We can add it easy later as model constant or static method.

For now it's looks little redundant.
Do you have business cases where this is useful per model?

File: dj_cqrs/utils.py (hunk 7 above)
Comment (net-free) [reply dj_cqrs/utils.py:18]: Thought default settings in constants it's not the best solution.
But if you ok, i will ok too.

File: requirements/dev.txt (hunk 9 above)
Comment (net-free) [reply requirements/dev.txt:6]: It's used for parsing expires string (from_message) to datetime with timezone info.
In python 3.6 i didn't found good solution,  datetime.fromisoformat only for python >= 3.7.
At summary:
- dateutil already at connect dependencies
- it's pretty stable
- 2.4 released at 2015 and should't broke any library dependencies

File: dj_cqrs/utils.py (hunk 8 above)
Comment (net-free) [reply dj_cqrs/utils.py:12]: For me current solution is simplest.

We could add special symbol to describe infinite if you don't like null/None for that. But it will dramatically increase code complexity (message body serialising and deserialising, comparison and etc. ).

Another open question is what is 'unreachable', we don't know at which moment this date will be compared.

File: dj_cqrs/delay.py (hunk 1 above)
Comment (net-free) [reply dj_cqrs/delay.py:29]: No, we always have comparable eta timestamp for priority, custom put(delay_message) guaranties that.

File: dj_cqrs/mixins.py (hunk 3 above)
Comment (net-free) [reply dj_cqrs/mixins.py:438]: max_retries = 3
| attempt        | current_retry           | should retry?  |
|-------------|-------------|-----|
| 1      | 0 | 0 < 3 = True |
| 2      | 1      |   2 < 3 = True |
| 3 | 2      |    2 < 3 = True |
| 4 | 3      |    3 < 3 = False |

total attempts = first + retries

File: dj_cqrs/signals.py (hunk 5 above)
Comment (maxipavlovic) [reply dj_cqrs/signals.py:76]: I don't (a the moment). But I don't have business cases, where it's needed per service. It has less sense, imho. So let's leave as is :)

File: requirements/dev.txt (hunk 9 above)
Comment (maxipavlovic) [reply requirements/dev.txt:6]: Okay, let's leave.
I just thought it was there in Django: 
https://docs.djangoproject.com/en/2.2/_modules/django/utils/dateparse/

But I don't know about version 3.
